### PR TITLE
Cake restructure

### PR DESCRIPTION
**Pull request by kroenlein**
Reworked Cake organization so that it might serve as the origin for tables.  Added ingredient amounts and some provenance.

### CODE REVIEW

File: taurus/demo/cake.py [hunk 1]
```diff
@@ -25,148 +26,540 @@
...
+        ],
+        notes='Chocolate frosting'
+    )
+    IngredientSpec(
+        name="{} input".format(frosting.name),
+        tags=list(frosting.tags),
+        labels=['coating'],
+        process=cake.process,
+        material=frosting,
+        absolute_quantity=NominalReal(nominal=1.0208, units='kg')
```
Comment (bfolie) [inline taurus/demo/cake.py]: Whoa that's a lot of frosting

File: taurus/entity/tests/test_link_by_uid.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    leaf = MaterialRun(name='leaf', process=ProcessRun(name='leaf proc'))
+    IngredientRun(process=root.process, material=leaf)
+    IngredientRun(process=root.process, material=LinkByUID.from_entity(leaf))
+
+    hist = complete_material_history(root)
+    copy = loads(dumps(hist))[-1]
+    assert copy.process.ingredients[0].material == copy.process.ingredients[1].material
+
+
+if __name__ == '__main__':
```
Comment (bfolie) [inline taurus/entity/tests/test_link_by_uid.py]: Leftover from debugging?

Comment (kroenlein) [reply taurus/entity/tests/test_link_by_uid.py]: Yes, good catch.

File: taurus/entity/tests/test_link_by_uid.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+from taurus.entity.util import complete_material_history
+from taurus.entity.object.material_run import MaterialRun
+from taurus.entity.object.process_run import ProcessRun
+from taurus.entity.object.ingredient_run import IngredientRun
+from taurus.entity.link_by_uid import LinkByUID
+
+
+def test_link_by_uid():
+    """Test that linking works."""
+
```
Comment (bfolie) [inline taurus/entity/tests/test_link_by_uid.py]:
```suggestion
```

**Pull request merged**